### PR TITLE
Add Hinge, Huber and LogCosh losses

### PR DESCRIPTION
**Pull request by nsantavas**
## Proposed changes

Add new loss functions: hinge_loss, huber_loss, and log_cosh_loss

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Comment by emircanerol** [comment]
You may want to see Awni's comments on my PR [Triplet Loss](https://github.com/ml-explore/mlx/pull/211):

- Replacing mx.array with array
- Maybe adding .. math:: section to each function for reports
- Testing loss functions with other reduction styles

**Comment by nsantavas** [comment]
@emircanerol @awni Done :)

**Comment by awni** [comment]
Thanks a ton for adding this! I will review soon!

File: python/mlx/nn/losses.py [hunk 1]
```diff
@@ -283,3 +283,83 @@ def _reduce(loss: mx.array, reduction: str = "none"):
...
+
+def huber_loss(
+    inputs: mx.array, targets: mx.array, delta: float = 1.0, reduction: str = "none"
+) -> mx.array:
+    """
+    Computes the Huber loss between inputs and targets.
+
+    .. math::
+
+       L_{\delta}(a) = \left\{ \begin{array}{ll}\frac{1}{2} a^2 & \text{for } |a| \leq \delta, \\\delta \left( |a| - \frac{1}{2} \delta \right) & \text{otherwise.}\end{array} \right.
```
Comment (angeloskath) [inline python/mlx/nn/losses.py:321]: Could you maybe format the latex in multiple lines? Also, I didn't render it but I think there is a brace missing after the last `\right`.

File: python/mlx/nn/losses.py [hunk 2]
```diff
@@ -283,3 +283,83 @@ def _reduce(loss: mx.array, reduction: str = "none"):
...
+
+def log_cosh_loss(
+    inputs: mx.array, targets: mx.array, reduction: str = "none"
+) -> mx.array:
+    """
+    Computes the log cosh loss between inputs and targets.
+
+    .. math::
+
+       \text{logcosh}(y_{\text{true}}, y_{\text{pred}}) = \frac{1}{n} \sum_{i=1}^{n} \log(\cosh(y_{\text{pred}}^{(i)} - y_{\text{true}}^{(i)}))
```
Comment (angeloskath) [inline python/mlx/nn/losses.py:351]: Same as above.

File: python/mlx/nn/losses.py [hunk 3]
```diff
@@ -283,3 +283,83 @@ def _reduce(loss: mx.array, reduction: str = "none"):
...
+        inputs (array): The predicted values.
+        targets (array): The target values.
+        reduction (str, optional): Specifies the reduction to apply to the output:
+          ``'none'`` | ``'mean'`` | ``'sum'``. Default: ``'none'``.
+
+    Returns:
+        array: The computed log cosh loss.
+    """
+    errors = inputs - targets
+    loss = mx.logaddexp(errors, -errors) - mx.log(mx.array(2.0))
```
Comment (angeloskath) [inline python/mlx/nn/losses.py:363]: I am a bit torn about the `mx.log(mx.array(2.0))`. It is totally fine to do that but for some reason I want this to be like a python constant. ie `math.log(2)` or even `0.6931471805599453 # log(2)` .

Feel free to do as you please with that one. Maybe @awni feels the opposite of me :-)

File: python/mlx/nn/losses.py (hunk 2 above)
Comment (angeloskath) [reply python/mlx/nn/losses.py:351]: Also feel free to drop a line about why this makes sense to be here as a loss function. Something like "log cosh behaves similar to error**2 for small values but |error| for large values which makes it robust to occasional outliers".

**Review by angeloskath — CHANGES_REQUESTED** [review]
This looks great! Thank you.

I left a couple style related comments. Could you please also add the new loss functions to `docs/src/python/nn/losses.rst`?

File: python/mlx/nn/losses.py (hunk 1 above)
Comment (nsantavas) [reply python/mlx/nn/losses.py:321]: Hey @angeloskath that's the rendered. I think it's ok

Screenshot 2023-12-21 at 09 42 19

File: python/mlx/nn/losses.py (hunk 2 above)
Comment (nsantavas) [reply python/mlx/nn/losses.py:351]: done

File: python/mlx/nn/losses.py (hunk 3 above)
Comment (nsantavas) [reply python/mlx/nn/losses.py:363]: done

File: python/mlx/nn/losses.py (hunk 1 above)
Comment (angeloskath) [reply python/mlx/nn/losses.py]: My bad, it looks great. Too used to seeing `\left( \right)` I jumped the gun :-)

**Review by angeloskath — CHANGES_REQUESTED** [review]
This looks perfect. Could you please also add the losses to the docs ?

File: python/mlx/nn/losses.py (hunk 3 above)
Comment (awni) [reply python/mlx/nn/losses.py]: +1 for constant here.

**Comment by nsantavas** [comment]
> This looks perfect. Could you please also add the losses to the docs ?

Done, also changed `mx.log(2)` to `math.log(2)` as @awni suggested :)

**Review by awni — APPROVED** [review]
Very nice!! I left two nits in the docstrings, could you please update and then we can merge?

File: python/mlx/nn/losses.py [hunk 4]
```diff
@@ -283,3 +285,94 @@ def _reduce(loss: mx.array, reduction: str = "none"):
...
+    Computes the hinge loss between inputs and targets.
+
+    .. math::
+
+       \text{hinge}(y, y_{\text{pred}}) = \max(0, 1 - y \cdot y_{\text{pred}})
+
+
+    Args:
+        inputs (array): The predicted values.
+        targets (array): The target values.
```
[the remark this answers is not in the document]
Comment (nsantavas) [reply python/mlx/nn/losses.py]: done

File: python/mlx/nn/losses.py [hunk 5]
```diff
@@ -283,3 +285,94 @@ def _reduce(loss: mx.array, reduction: str = "none"):
...
+
+
+def log_cosh_loss(
+    inputs: mx.array, targets: mx.array, reduction: str = "none"
+) -> mx.array:
+    """
+    Computes the log cosh loss between inputs and targets.
+
+    Logcosh acts like L2 loss for small errors, ensuring stable gradients,
+    and like L1 loss for large errors, reducing sensitivity to outliers. This
```
[the remark this answers is not in the document]
Comment (nsantavas) [reply python/mlx/nn/losses.py]: done

**Comment by awni** [comment]
Perfect! I thnk this can be merged as soon as @angeloskath approves !

**Review by angeloskath — APPROVED** [review]
Super nice :-)
